src/decoder/error.rs

```diff
@@ -1,39 +1,60 @@
 use std::{error, fmt, fmt::Display, io, string};
 
-use serde::de::{self, Expected, Unexpected};
+use crate::Bson;
+use de::Unexpected;
+use serde::de;
 
 /// Possible errors that can arise during decoding.
 #[derive(Debug)]
 #[non_exhaustive]
 pub enum DecoderError {
+    /// A `std::io::Error` encountered while deserializing.
     IoError(io::Error),
+
+    /// A `std::string::FromUtf8Error` encountered while decoding a UTF-8 String from the input
+    /// data.
     FromUtf8Error(string::FromUtf8Error),
-    UnrecognizedDocumentElementType { key: String, element_type: u8 },
-    InvalidArrayKey(usize, String),
-    // A field was expected, but none was found.
-    ExpectedField(&'static str),
-    // An unexpected field was found.
-    UnknownField(String),
-    // There was an error with the syntactical structure of the BSON.
-    SyntaxError(String),
-    // The end of the BSON input was reached too soon.
+
+    /// While decoding a `Document` from bytes, an unexpected or unsupported element type was
+    /// encountered.
+    UnrecognizedDocumentElementType {
+        /// The key at which an unexpected/unsupported element type was encountered.
+        key: String,
+
+        /// The encountered element type.
+        element_type: u8,
+    },
+
+    /// There was an error with the syntactical structure of the BSON.
+    SyntaxError { message: String },
+
+    /// The end of the BSON input was reached too soon.
     EndOfStream,
-    // Invalid Type
```

Comment:
Removing these fields is part of the work of RUST-444. This ticket hadn't gone through triage yet, but seeing as the clock is ticking down I just went ahead and implemented it. I can always revert it if we end up thinking it wasn't the right idea, but I figured it was unlikely that would happen.